src/main/java/org/ice4j/util/AsyncQueueHandler.java

```diff
@@ -0,0 +1,261 @@
+                        return;
+                    }
+                }
+
+                sequentiallyHandledItems++;
+
+                try
+                {
+                    handler.handleItem(item);
+                }
+                catch (Throwable e)
+                {
+                    logger.warning("Failed to handle item: " + e);
+                }
+            }
+        }
+    };
+
+    /**
+     * Constucts instance of {@link AsyncQueueHandler<T>} which is capable of
+     * asyncronous reading provided queue from thread borrowed from executor to
+     * process items with provided handler.
+     * @param queue thread-safe queue which holds items to process
+     * @param handler an implementation of handler routine which will be
+     * invoked per each item placed in the queue.
+     * @param id optional identifier of current handler for debug purpose
+     * @param executor optional executor service to borrow threads from
+     * @param maxSequentiallyHandledItems maximum number of items sequentially
+     * handled on thread borrowed from {@link #executor} before temporary
+     * releasing thread and re-acquiring it from {@link #executor}.
+     */
+    public AsyncQueueHandler(
+        BlockingQueue<T> queue,
+        Handler<T> handler,
+        String id,
+        ExecutorService executor,
+        long maxSequentiallyHandledItems)
+    {
+        if (queue == null)
+        {
+            throw new IllegalArgumentException("queue is null");
+        }
+
+        if (handler == null)
+        {
+            throw new IllegalArgumentException("handler is null");
+        }
+
+        this.executor = executor != null ? executor : sharedExecutor;
+        this.queue = queue;
+        this.handler = handler;
+        this.id = id;
+        this.maxSequentiallyHandledItems = maxSequentiallyHandledItems;
+    }
+
+    /**
+     * Attempts to stop execution of {@link #reader} if running
+     */
+    public void cancel()
+    {
+        running.set(false);
+        cancel(true);
+    }
+
+    /**
+     * Checks if {@link #reader} is running on one of {@link #executor}
+     * thread and if no submits execution of {@link #reader} on executor.
+     */
+    public void handleQueueItemsUntilEmpty()
+    {
+        synchronized (syncRoot)
+        {
+            if (readerFuture == null || readerFuture.isDone())
+            {
+                running.set(true);
+                readerFuture = executor.submit(reader);
+            }
+        }
+    }
+
+    /**
+     * Invoked when execution of {@link #reader} is about to temporary
+     * cancel and further execution need to be re-scheduled.
+     * Assuming called when lock on {@link #syncRoot} is already taken.
+     */
+    private void onYield()
+    {
+        if (logger.isLoggable(java.util.logging.Level.FINE))
+        {
+            logger.fine("Yielding AsyncQueueHandler associated with "
+                + "AsyncQueueHandler with ID = " + id);
```

Comment:
should this log just be `"Yielding AsyncQueueHandler associated with ID " + id`?